### PR TITLE
net: lib: nrf_cloud: allow runtime client id for coap

### PR DESCRIPTION
**Pull request by jayteemo**
Add function nrf_cloud_client_id_runtime_set.
Allow runtime client ID for CoAP and REST.

IRIS-9081

### CODE REVIEW

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -400,7 +400,15 @@ Multiprotocol Service Layer libraries
...
-|no_changes_yet_note|
+* :ref:`lib_nrf_cloud` library:
+
+  * Added:
+
+    * The :c:func:`nrf_cloud_client_id_runtime_set` function to set the device ID string if the :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_SRC_RUNTIME` Kconfig option is enabled.
+
+  * Updated:
+
+    * The :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_SRC_RUNTIME` Kconfig can now be used with CoAP and REST.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:411]:
```suggestion
  * Added the :c:func:`nrf_cloud_client_id_runtime_set` function to set the device ID string if the :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_SRC_RUNTIME` Kconfig option is enabled.

  * Updated the :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_SRC_RUNTIME` Kconfig option to be available with CoAP and REST.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: include/net/nrf_cloud.h [hunk 2]
```diff
@@ -822,13 +822,30 @@ int nrf_cloud_modem_fota_completed(const bool fota_success);
 /**
  * @brief Retrieve the current device ID.
  *
- * @param[in,out] id_buf Buffer to receive the device ID.
- * @param[in] id_len     Size of buffer (NRF_CLOUD_CLIENT_ID_MAX_LEN).
+ * @param[in,out] id_buf Buffer to receive the device ID as a null-terminated string.
+ * @param[in] id_len     Size of buffer, maximum size is NRF_CLOUD_CLIENT_ID_MAX_LEN + 1.
```
Comment (peknis) [inline include/net/nrf_cloud.h:826]: Cannot seem to find this one below. Might be this failing the doc build.

File: include/net/nrf_cloud.h [hunk 3]
```diff
@@ -822,13 +822,30 @@ int nrf_cloud_modem_fota_completed(const bool fota_success);
...
+ *        Requires @kconfig{CONFIG_NRF_CLOUD_CLIENT_ID_SRC_RUNTIME} to be enabled.
+ *
+ * @param[in] client_id Null-terminated device ID string.
+ *                      Max string length is NRF_CLOUD_CLIENT_ID_MAX_LEN.
  *
  * @retval 0 If successful.
  * @return A negative value indicates an error.
  */
-int nrf_cloud_client_id_get(char *id_buf, size_t id_len);
+int nrf_cloud_client_id_runtime_set(const char *const client_id);
```
Comment (peknis) [inline include/net/nrf_cloud.h:848]: Or is it this one failing the build. Extra const?

**Review by glarsennordic — APPROVED** [review]
LGTM in general, just some suggested tweaks

**Review by peknis — APPROVED** [review]

**Review by plskeggs — CHANGES_REQUESTED** [review]

**Review by plskeggs — APPROVED** [review]

File: include/net/nrf_cloud.h [hunk 4]
```diff
@@ -822,13 +822,30 @@ int nrf_cloud_modem_fota_completed(const bool fota_success);
...
+ *                   but the runtime client ID has not been set.
+ *                   See @ref nrf_cloud_client_id_runtime_set.
+ * @return A negative value indicates an error.
+ */
+int nrf_cloud_client_id_get(char *id_buf, size_t id_buf_sz);
+
+/**
+ * @brief Set the device ID at runtime.
+ *        Requires @kconfig{CONFIG_NRF_CLOUD_CLIENT_ID_SRC_RUNTIME} to be enabled.
+ *
```
Comment (plskeggs) [inline include/net/nrf_cloud.h:841]: We should provide some guidance either here or in the library docs that explains it is up to the caller to manage the network connection when changing the device id. The device will fail to communicate with nrfcloud using this device id if it is not already onboarded / registered using this device id.

Comment (jayteemo) [reply include/net/nrf_cloud.h:841]: > The device will fail to communicate with nrfcloud using this device id if it is not already onboarded / registered using this device id.

this is true for any `CONFIG_NRF_CLOUD_CLIENT_ID_SRC_` that is selected.
you already have a ticket for this doc improvement 😄 
IRIS-5868

> it is up to the caller to manage the network connection when changing the device id.

added a note

**Review by plskeggs — APPROVED** [review]

**Pull request merged**